lib/ASTGen/Sources/ASTGen/Attrs.swift

```diff
@@ -0,0 +1,1101 @@
+      name: name
+    )
+  }
+
+  func generateOptimizeAttr(attribute node: AttributeSyntax) -> BridgedOptimizeAttr? {
+    let mode: BridgedOptimizationMode? = self.generateSingleAttrOption(
+      attribute: node,
+      {
+        switch $0.rawText {
+        case "speed": return .forSpeed
+        case "size": return .forSize
+        case "none": return .noOptimization
+        default: return nil
+        }
+      }
+    )
+    guard let mode else {
+      return nil
+    }
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      range: self.generateSourceRange(node),
+      mode: mode
+    )
+  }
+
+  func generateOriginallyDefinedInAttr(attribute node: AttributeSyntax) -> [BridgedOriginallyDefinedInAttr] {
+    guard
+      // `@_OriginallyDefinedIn` has special argument list syntax.
+      let args = node.arguments?.as(OriginallyDefinedInAttributeArgumentsSyntax.self)
+    else {
+      // TODO: Diagnose.
+      return []
+    }
+
+    fatalError("unimplemented")
+  }
+
+  func generatePrivateImportAttr(attribute node: AttributeSyntax) -> BridgedPrivateImportAttr? {
+    guard
+      // `@_private` has special argument list syntax
+      let args = node.arguments?.as(UnderscorePrivateAttributeArgumentsSyntax.self)
+    else {
+      // TODO: Diagnose
+      return nil
+    }
+
+    guard let fileName = self.generateStringLiteralTextIfNotInterpolated(expr: args.filename) else {
+      // TODO: Diagnose
+      return nil
+    }
+
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      attrNameLoc: self.generateSourceLoc(node.attributeName),
+      lParenLoc: self.generateSourceLoc(node.leftParen),
+      fileName: fileName,
+      rParenLoc: self.generateSourceLoc(node.rightParen)
+    )
+  }
+
+  func generateProjectedValuePropertyAttr(attribute node: AttributeSyntax) -> BridgedProjectedValuePropertyAttr? {
+    // `@_dynamicReplacement` has special argument list syntax
+    let name = self.generateSingleAttrOption(attribute: node, { self.generateIdentifier($0) }, valueIfOmitted: BridgedIdentifier())
+    guard let name else {
+      // TODO: Diagnose.
+      return nil
+    }
+
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      range: self.generateSourceRange(node),
+      name: name
+    )
+  }
+
+  func generateReferenceOwnershipAttr(attribute node: AttributeSyntax, attrName: SyntaxText)
+    -> BridgedReferenceOwnershipAttr?
+  {
+    let kind: BridgedReferenceOwnership
+    if attrName == "weak" {
+      kind = .weak
+    } else {
+      kind =
+        self.generateSingleAttrOption(
+          attribute: node,
+          {
+            switch $0.rawText {
+            case "unsafe": return .unmanaged
+            case "safe": return .unowned
+            default: return nil
+            }
+          },
+          valueIfOmitted: .unowned
+        ) ?? .unowned
+    }
```

Comment:
Should we be diagnosing if it's not named `weak` or `unowned`?

Reply:
Also do we need to pass `valueIfOmitted` if we're `??`'ing the result anyway?

Reply:
`valueIfOmitted` is used when there's no argument clause. If we don't pass it, `generateSingleAttrOption` diagnose it as "missing argument clause".

Reply:
Ah I see, I missed the TODO diagnostic